### ISSUE TITLE
Arduino Nano ESP32

### ISSUE THREAD

**Issue by PeterUMJ** [mrrwa/LocoNet2#12]
Hi

I have tried with Arduino Nano ESP32, get the error Unsupported ESP32 Archutecture.
From what I can see in the LocoNetStreamESP32.h file, this is what is causing the error.

#if defined(ARDUINO_ESP32_MCU_esp32c3)    
	return hw->fsm_status.st_urx_out != 0;
#elif defined(ARDUINO_ESP32_MCU_esp32)  
        return hw->status.st_urx_out != 0;
#else
#error "Unsupported ESP32 Architecture"
#endif
};

Nano has ESP32-S3.

 have tried changing the file and adding that type, but not working. I don't know what needs to be done, except to change this file.
However, tried putting this line in else "return hw->fsm_status.st_urx_out != 0";
Tried a bit and it worked, guess the difference between esp32c3 and esp32s3 doesn't matter maybe?
Thanks in advance if it can be fixed.

Best Regards
Peter

**Comment by kiwi64ajs** [comment]
Hi Peter,
I've made a few changes to the LocoNetStreamESP32.h file in the library to print a better error message as I had to add a custom change to the ESP32 "platform.txt" file to allow for some conditional compilation to handle ESP32 'mcu' specific differences. Maybe there is a better way to do this, but it is all I could think of... ;) 

Essentially, you need to edit the platform.txt file under the ESP32 Arduino core and search for
    '-DARDUINO_PARTITION_{build.partitions}'
 and make sure it is followed by
     '-DARDUINO_ESP32_MCU_{build.mcu}'
If it's not, then add it.

I really need to fork the ESP32 Arduino core repo and submit a PR that has this change... ;)

**Comment by PeterUMJ** [comment]
Hi 

I have changed as you suggest, hope it is the right file, this one

"C:\Users\Peter\AppData\Local\Arduino15\packages\arduino\hardware\esp32\2.0.18-20240930.arduino3\platform.txt"

Changed the following:

## Compile c files
recipe.c.o.pattern="{compiler.path}{compiler.c.cmd}" {compiler.cpreprocessor.flags} {compiler.c.flags} -DF_CPU={build.f_cpu} -DARDUINO={runtime.ide.version } -DARDUINO_{build.board} -DARDUINO_ARCH_{build.arch} -DARDUINO_BOARD="{build.board}" -DARDUINO_VARIANT="{build.variant}" -DARDUINO_PARTITION_{build.partitions}
-DARDUINO_ESP32_MCU_{build.mcu} {compiler.c.extra_flags} {build.extra_flags} "@{build.opt.path}" "@{file_opts.path}" {includes} "{source_file}" -o "{object_file }"

## Compile c++ files
recipe.cpp.o.pattern="{compiler.path}{compiler.cpp.cmd}" {compiler.cpreprocessor.flags} {compiler.cpp.flags} -DF_CPU={build.f_cpu} -DARDUINO={runtime. ide.version} -DARDUINO_{build.board} -DARDUINO_ARCH_{build.arch} -DARDUINO_BOARD="{build.board}" -DARDUINO_VARIANT="{build.variant}" -DARDUINO_PARTITION_{build.partitions}
-DARDUINO_ESP32_MCU_{build.mcu} {compiler.cpp.extra_flags} {build.extra_flags} "@{build.opt.path}" "@{file_opts.path}" {includes} "{source_file}" -o "{object_file }"

## Compile S files
recipe.S.o.pattern="{compiler.path}{compiler.c.cmd}" {compiler.cpreprocessor.flags} {compiler.S.flags} -DF_CPU={build.f_cpu} -DARDUINO={runtime.ide.version } -DARDUINO_{build.board} -DARDUINO_ARCH_{build.arch} -DARDUINO_BOARD="{build.board}" -DARDUINO_VARIANT="{build.variant}" -DARDUINO_PARTITION_{build.partitions}
-DARDUINO_ESP32_MCU_{build.mcu} {compiler.S.extra_flags} {build.extra_flags} "@{build.opt.path}" "@{file_opts.path}" {includes} "{source_file}" -o "{object_file }"

Then I get the following error:
Invalid FQBN: board arduino:esp32:nano_nora not found

Peter

**Comment by kiwi64ajs** [comment]
Hi Peter,

The changes to the platform.txt file look ok, but that error is odd. I was only seeing compile time errors.

I’ve got core version 2.0.17, and I noticed you have version 2.0.18, so maybe I need to try that.

Also, I changed the file: https://github.com/mrrwa/LocoNet2/blob/development/src/LocoNetStreamESP32.h , specifically to the issue you had. So, you will need both the platform.txt file change and the ESP32 library support file change.

HTH

Alex

**Comment by PeterUMJ** [comment]
Hi Alex

New LocoNetStreamESP32.h and changed platform.txt

I still get 
"Unknown FQBN: board arduino:esp32:nano_nora not found. Could not connect to COM4 serial port"
when open ide.

New LocoNetStreamESP32.h and old platform.txt I get this when compiling:

"Unsupported ESP32 Archutecture. Check the ESP32 platform.txt file to ensure every occurance of '-DARDUINO_PARTITION_{build.partitions}' is followed by '-DARDUINO_ESP32_MCU_{build.mcu}'

Could I have corrupted the platform.txt file when I changed it, maybe?

Best Regards
Peter

**Comment by kiwi64ajs** [comment]
Hi Peter,

> On 26 Oct 2024, at 5:13 AM, PeterUMJ ***@***.***> wrote:
> Hi Alex
> New LocoNetStreamESP32.h and changed platform.txt
> 
> I still get
> "Unknown FQBN: board arduino:esp32:nano_nora not found. Could not connect to COM4 serial port"
> when open ide.
> 
Your issue may now be unrelated to your initial problem, which the changes I made fixed.

See: https://support.arduino.cc/hc/en-us/articles/4412950941074-Compilation-error-Missing-FQBN-Fully-Qualified-Board-Name

I just tried the code again on the 3.1.0-RC2 core, editing the platform.txt file as before, and the code compiles fine.

Ensure you can compile the Blink example to ensure your Arduino Dev environment is working correctly before trying the LocoNet2 Basic-Listener-ESP32.ino example.

Regards

Alex Shepherd

**Comment by PeterUMJ** [comment]
Hi Alex

I'm away this week, I'll check next week when I'm back home

Regards
Peter

**Comment by PeterUMJ** [comment]
Hi Alex

I don't get the error Invalid FQBN: board arduino:esp32:nano_nora not found anymore. I make a new platform.txt, I
probably got some junk when I edited it the first time.

But when I compile the sketch Basic_Listener I get this, the Blink example work fine. I have both LocoNet and LocoNet2 installed, is that a problem?

In file included from C:\Users\Peter\Documents\Arduino\libraries\Embedded_Template_Library_ETL\src/etl/array.h:44,
                 from C:\Users\Peter\Documents\Arduino\libraries\Embedded_Template_Library_ETL\src/etl/vector.h:42,
                 from C:\Users\Peter\Documents\Arduino\libraries\LocoNet2\src/LocoNet2.h:72,
                 from C:\Users\Peter\Documents\Arduino\libraries\LocoNet2\src/LocoNetStream.h:42,
                 from C:\Users\Peter\Documents\Arduino\libraries\LocoNet2\src/LocoNetStreamESP32.h:40,
                 from Z:\Arduino LocoNet2\Basic_Listener\Basic_Listener.ino:5:
C:\Users\Peter\Documents\Arduino\libraries\Embedded_Template_Library_ETL\src/etl/initializer_list.h:143:9: error: redefinition of 'class std::initializer_list<_E>'
   class initializer_list
         ^~~~~~~~~~~~~~~~
In file included from c:\users\peter\appdata\local\arduino15\packages\esp32\tools\xtensa-esp32s3-elf-gcc\esp-2021r2-patch5-8.4.0\xtensa-esp32s3-elf\include\c++\8.4.0\utility:76,
                 from c:\users\peter\appdata\local\arduino15\packages\esp32\tools\xtensa-esp32s3-elf-gcc\esp-2021r2-patch5-8.4.0\xtensa-esp32s3-elf\include\c++\8.4.0\algorithm:60,
                 from C:\Users\Peter\AppData\Local\Arduino15\packages\arduino\hardware\esp32\2.0.18-20240930.arduino3\cores\esp32/Arduino.h:184,
                 from sketch\Basic_Listener.ino.cpp:1:
c:\users\peter\appdata\local\arduino15\packages\esp32\tools\xtensa-esp32s3-elf-gcc\esp-2021r2-patch5-8.4.0\xtensa-esp32s3-elf\include\c++\8.4.0\initializer_list:47:11: note: previous definition of 'class std::initializer_list<_E>'
     class initializer_list
           ^~~~~~~~~~~~~~~~
In file included from C:\Users\Peter\Documents\Arduino\libraries\Embedded_Template_Library_ETL\src/etl/array.h:44,
                 from C:\Users\Peter\Documents\Arduino\libraries\Embedded_Template_Library_ETL\src/etl/vector.h:42,
                 from C:\Users\Peter\Documents\Arduino\libraries\LocoNet2\src/LocoNet2.h:72,
                 from C:\Users\Peter\Documents\Arduino\libraries\LocoNet2\src/LocoNetStream.h:42,
                 from C:\Users\Peter\Documents\Arduino\libraries\LocoNet2\src/LocoNetStreamESP32.h:40,
                 from Z:\Arduino LocoNet2\Basic_Listener\Basic_Listener.ino:5:
C:\Users\Peter\Documents\Arduino\libraries\Embedded_Template_Library_ETL\src/etl/initializer_list.h:205:22: error: redefinition of 'template<class T> constexpr const T* std::begin(std::initializer_list<_Tp>)'
   constexpr const T* begin(initializer_list<T> init) noexcept
                      ^~~~~
In file included from c:\users\peter\appdata\local\arduino15\packages\esp32\tools\xtensa-esp32s3-elf-gcc\esp-2021r2-patch5-8.4.0\xtensa-esp32s3-elf\include\c++\8.4.0\utility:76,
                 from c:\users\peter\appdata\local\arduino15\packages\esp32\tools\xtensa-esp32s3-elf-gcc\esp-2021r2-patch5-8.4.0\xtensa-esp32s3-elf\include\c++\8.4.0\algorithm:60,
                 from C:\Users\Peter\AppData\Local\Arduino15\packages\arduino\hardware\esp32\2.0.18-20240930.arduino3\cores\esp32/Arduino.h:184,
                 from sketch\Basic_Listener.ino.cpp:1:
c:\users\peter\appdata\local\arduino15\packages\esp32\tools\xtensa-esp32s3-elf-gcc\esp-2021r2-patch5-8.4.0\xtensa-esp32s3-elf\include\c++\8.4.0\initializer_list:89:5: note: 'template<class _Tp> constexpr const _Tp* std::begin(std::initializer_list<_Tp>)' previously declared here
     begin(initializer_list<_Tp> __ils) noexcept
     ^~~~~
In file included from C:\Users\Peter\Documents\Arduino\libraries\Embedded_Template_Library_ETL\src/etl/array.h:44,
                 from C:\Users\Peter\Documents\Arduino\libraries\Embedded_Template_Library_ETL\src/etl/vector.h:42,
                 from C:\Users\Peter\Documents\Arduino\libraries\LocoNet2\src/LocoNet2.h:72,
                 from C:\Users\Peter\Documents\Arduino\libraries\LocoNet2\src/LocoNetStream.h:42,
                 from C:\Users\Peter\Documents\Arduino\libraries\LocoNet2\src/LocoNetStreamESP32.h:40,
                 from Z:\Arduino LocoNet2\Basic_Listener\Basic_Listener.ino:5:
C:\Users\Peter\Documents\Arduino\libraries\Embedded_Template_Library_ETL\src/etl/initializer_list.h:214:22: error: redefinition of 'template<class T> constexpr const T* std::end(std::initializer_list<_Tp>)'
   constexpr const T* end(initializer_list<T> init) noexcept
                      ^~~
In file included from c:\users\peter\appdata\local\arduino15\packages\esp32\tools\xtensa-esp32s3-elf-gcc\esp-2021r2-patch5-8.4.0\xtensa-esp32s3-elf\include\c++\8.4.0\utility:76,
                 from c:\users\peter\appdata\local\arduino15\packages\esp32\tools\xtensa-esp32s3-elf-gcc\esp-2021r2-patch5-8.4.0\xtensa-esp32s3-elf\include\c++\8.4.0\algorithm:60,
                 from C:\Users\Peter\AppData\Local\Arduino15\packages\arduino\hardware\esp32\2.0.18-20240930.arduino3\cores\esp32/Arduino.h:184,
                 from sketch\Basic_Listener.ino.cpp:1:
c:\users\peter\appdata\local\arduino15\packages\esp32\tools\xtensa-esp32s3-elf-gcc\esp-2021r2-patch5-8.4.0\xtensa-esp32s3-elf\include\c++\8.4.0\initializer_list:99:5: note: 'template<class _Tp> constexpr const _Tp* std::end(std::initializer_list<_Tp>)' previously declared here
     end(initializer_list<_Tp> __ils) noexcept
     ^~~

exit status 1
Compilation error: exit status 1



Regards
Peter

**Comment by PeterUMJ** [comment]
Hi Alex

I removed the Loconet2 library and reinstalled it. Now compiling Basic_Listener worked.
I haven't tested it any further.

If you are not aware of it. then Loconet2 does not appear in the Library Manager.

Thanks for your help

Regards
Peter

**Comment by kiwi64ajs** [comment]
Hi Peter,That’s good news. Before listing the LocoNet2 library in the Library Manager I was waiting for a chance to check/compare/review the code and subsequent changes/bug fixes that were done in the LocoNet library after I took the LocoNet code and refactored it into LocoNet2.Once I’m satisfied that it’s not behind the LocoNet library, then I’ll register it.It would also be good to have the LocoNet2 library also support the bit-bashed AVR  interface on the AVRs. RegardsAlex ShepherdOn 7 Nov 2024, at 10:14 PM, PeterUMJ ***@***.***> wrote:﻿
Hi Alex
I removed the Loconet2 library and reinstalled it. Now compiling Basic_Listener worked.
I haven't tested it any further.
If you are not aware of it. then Loconet2 does not appear in the Library Manager.
Thanks for your help
Regards
Peter
